frontend/src/metabase/dashboard/components/DashboardEmbedAction/DashboardEmbedAction.tsx

```diff
@@ -0,0 +1,116 @@
+import { useState } from "react";
+import type { Dashboard } from "metabase-types/api";
+import { DashboardEmbedHeaderButton } from "metabase/dashboard/components/DashboardEmbedHeaderButton";
+import { DashboardEmbedHeaderMenu } from "metabase/dashboard/components/DashboardEmbedHeaderMenu";
+import DashboardSharingEmbeddingModal from "metabase/dashboard/containers/DashboardSharingEmbeddingModal";
+import { useSelector } from "metabase/lib/redux";
+import { getSetting } from "metabase/selectors/settings";
+import { getUserIsAdmin } from "metabase/selectors/user";
+
+export type EmbedButtonClickBehavior =
+  | "embed-menu"
+  | "embed-modal"
+  | "public-link-popover"
+  | null;
+
+const getClickBehavior = ({
+  isAdmin,
+  isPublicSharingEnabled,
+  hasPublicLink,
+}: {
+  isAdmin: boolean;
+  isPublicSharingEnabled: boolean;
+  hasPublicLink: boolean;
+}): EmbedButtonClickBehavior => {
+  if (!isPublicSharingEnabled) {
+    return isAdmin ? "embed-modal" : null;
+  }
+
+  if (isAdmin) {
+    return "embed-menu";
+  }
+
+  if (hasPublicLink) {
+    return "public-link-popover";
+  }
+
+  return null;
+};
+
+export const DashboardEmbedAction = ({
+  dashboard,
+}: {
+  dashboard: Dashboard;
+}) => {
+  const [isOpen, setIsOpen] = useState(false);
```

Comment:
Here we have 3 states that control how we want to show a menu or a modal
1. `isOpen`
2. `initialClickBehavior`
3. `clickBehavior`

It seems to me these could potentially be consolidated into a single state enum e.g.
- `null` (not opening anything)
- `embed-menu`
- `public-link-popover`
- `embed-modal`

And we could use this single state to determine which UI we want to show instead of using 3. Could you try doing this? I think if this is possible it would reduce the complexity of this component and make it a lot easier to understand.

Reply:
I was thinking about this as well - I was going to write a whole explanation about what I'm thinking and how we can work through it, but I figured it'll be faster to just try it out, see how it feels, and then go from there 😄